### ISSUE TITLE
 3rd party batch model (b3dm) in Mapbox v3? 

### ISSUE THREAD

**Issue by bertt** [mapbox/mapbox-gl-js#12861]
Hi, is there some information/minimal sample (without compression) of visualizing  a b3dm in Mapbox  v3?

questions:

 - What is the axis ordening? (z up? in  meters?)

- What coordinates to use in the glTF? epsg:3857? absolute or relative to the tile?

- How does it work when enabling terrain?

Any other requirements in the glTF? Draco compression? MAPBOX_mesh_features extension? mapbox:footprint:id/mapbox:coordinate-space/mapbox:footprint-quantized/mapbox:footprint-version/mapbox:tile-version?

**Comment by astojilj** [comment]
Thanks @bertt 

For the context, pasting earlier comment https://twitter.com/astojilj/status/1693498420579766656?s=20

> _Mapbox tiles use Z-up orientation with Z in meters and X and Y in tile coordinates https://github.com/mapbox/mapbox-gl-js-internal/blob/94c7b65cdc8a861eeddd12e4f726407984e8c879/3d-style/render/draw_model.js#L741. Please, use https://github.com/mapbox/mapbox-gl-js to request and then track progress on supporting 3rd party tiles that follow 3D Tiles 1.0 and 3D Tiles Next._

Note that this is experimental functionality, and we don't plan to support experiments with custom b3dm tiles while we are focusing on developing [Standard Style](https://www.mapbox.com/blog/standard-core-style). Our future plans include 3D Tiles Next support, but I cannot share any details now.

> What x, y coordinates to use in the glTF? epsg:3857? absolute or relative to the tile? Relative to what?

It follows Web Mercator projection (epsg:3857) - Mapbox Vector Tile Format https://docs.mapbox.com/style-spec/reference/sources/#vector.

> _All geometric coordinates in vector tiles must be between -1 * extent and (extent * 2) - 1 inclusive_

EXTENT: https://github.com/mapbox/mapbox-gl-js/blob/068a7f32954b3cb3d0754ec0e263b816a9eb3fd1/src/style-spec/data/extent.js#L18

**Comment by bertt** [comment]
Got a step further, see https://bertt.github.io/mapbox_3dtiles_samples/samples/standard/dam_square/ (MapBox version v3.1.2)

![image](https://github.com/mapbox/mapbox-gl-js/assets/538812/2c529015-f9d6-4c47-8d17-2aa0bca63822)

Loading tileset:

```
map.addSource('bag-3d', {
        "type": "batched-model",
        "maxzoom": 15,
        "minzoom": 15,
        "tiles": [
          "https://bertt.github.io/mapbox_3dtiles_samples/samples/standard/dam_square/tiles/{z}-{x}-{y}.glb"
        ]
      }
)});

// add the custom style layer to the map
map.on('style.load', () => {
  map.addLayer({
    id: 'bag-layer',
    type: 'model',
    source: 'bag-3d',          
  });
});
```

There is one glb (15-16829-10768.glb), center should be at Amsterdam Dam Square (
![image](https://github.com/mapbox/mapbox-gl-js/assets/538812/bc8ee9f2-796e-46bc-b2a3-5cb1921a45d8)
). In this demo I used in the vertex coordinates between 3500 and 3600 after trial and error, to get the model on the Dam square. When using 0,0 the model is drawn on the top left of the tile. So it's still unclear to me how the (x,y) coordinate system works and the relation to the extent (8192).

**Comment by astojilj** [comment]
@bertt 
Yes, the tile coordinate X increases from left to right in range [0 .. 8192), and Y increases [0 .. 8192) from north to south. Now, we use single zoom 14 .glb tiles only.

I would like to understand your use case better as we are working on a number of things:

https://github.com/mapbox/mapbox-gl-js/issues/13009#issuecomment-1900849976

Note the approaches for hiding all 3D content from Standard Style and the "conflation" part related to bringing own 3D models to the map, which would hide overlapping Mapbox 3D content.

**Comment by bertt** [comment]
@astojilj main use case for now is getting something similar as in build in CesiumJS: 

https://geodan.github.io/pg2b3dm/sample_data/3dbag/sibbe/

![Screenshot 2024-02-26 194916](https://github.com/mapbox/mapbox-gl-js/assets/538812/80673c38-a820-4725-bd73-2ad7db4f97fc)

Main features:

- Display 3D Tiles/glTF's

- Click on building to get attribute info (based on EXT_Structural_Metadata or b3dm header info)

- Combine with (custom) terrain, note we use 'depthTestAgainstTerrain' to occlude buildings behind hills  

- Nice to have: style 3D Tiles in the client based on attribute, for example style on building year

**Comment by bertt** [comment]
Delaware buildings https://bertt.github.io/mapbox_3dtiles_samples/samples/standard/delaware/ 
![image](https://github.com/mapbox/mapbox-gl-js/assets/538812/8e88e67d-57ba-4012-b21e-2344935d6dae)

**Comment by astojilj** [comment]
cc @jtorresfabra

**Comment by bertt** [comment]
q: Is there a sample of click on 3D Tiles and show attributes?

**Comment by bertt** [comment]
Dutch 3D Bag https://bertt.github.io/mapbox_3dtiles_samples/samples/standard/tienhoven/ 
![image](https://github.com/mapbox/mapbox-gl-js/assets/538812/d155db62-95fc-4795-b667-8fc0a3b200e8)

**Comment by astojilj** [comment]
@bertt 

Work on query render features support for Model Layer is in progress.

**Comment by enersis-pst** [comment]
@astojilj is there a official roadmap? would be nice to know what will come :-)

**Comment by spidunno** [comment]
> [@astojilj](https://github.com/astojilj) is there a official roadmap? would be nice to know what will come :-)

I'm also curious

**Comment by capriverse** [comment]
I noticed that MapLibre has updated its example for loading 3D Tiles, which is rendered based on the 3DTilesRendererJS plugin. Can Mapbox GL JS also render 3D Tiles in a similar way?
https://maplibre.org/maplibre-gl-js/docs/examples/add-3d-tiles-using-threejs/
